jib-core/src/test/java/com/google/cloud/tools/jib/cache/RetryTest.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.jib.cache;
+
+import java.util.concurrent.TimeUnit;
+import org.junit.Assert;
+import org.junit.Test;
+
+/** Tests for {@link Retry}. */
+public class RetryTest {
+  private int actionCount = 0;
+  private final Retry.Action<Exception> successfulAction =
+      () -> {
+        ++actionCount;
+        return true;
+      };
+  private final Retry.Action<Exception> unsuccessfulAction =
+      () -> {
+        ++actionCount;
+        return false;
+      };
+  private final Retry.Action<Exception> exceptionAction =
+      () -> {
+        ++actionCount;
+        throw new Exception("whee");
+      };
+
+  @Test
+  public void testSuccessfulAction() throws Exception {
+    boolean result = Retry.action(successfulAction).run();
+    Assert.assertTrue(result);
+    Assert.assertEquals(1, actionCount);
+  }
+
+  @Test
+  public void testMaximumRetries_default() throws Exception {
+    boolean result = Retry.action(unsuccessfulAction).run();
+    Assert.assertFalse(result);
+    Assert.assertEquals(5, actionCount);
+  }
+
+  @Test
+  public void testMaximumRetries_specified() throws Exception {
+    // if the stop condition is true then the action is never invoked
+    boolean result = Retry.action(unsuccessfulAction).maximumRetries(2).run();
+    Assert.assertFalse(result);
+    Assert.assertEquals(2, actionCount);
+  }
+
+  @Test
+  public void testRetryableException() {
+    // all exceptions are retryable by default, so retry 5 times
+    try {
+      Retry.action(exceptionAction).run();
+      Assert.fail("should have thrown exception");
+    } catch (Exception ex) {
+      Assert.assertEquals("whee", ex.getMessage());
+      Assert.assertEquals(5, actionCount);
+    }
+  }
+
+  @Test
+  public void testNonRetryableException() {
+    // the exception is not ok and so should only try 1 time
+    try {
+      Retry.action(exceptionAction).retryOnException(ex -> false).run();
+      Assert.fail("should have thrown exception");
+    } catch (Exception ex) {
+      Assert.assertEquals("whee", ex.getMessage());
+      Assert.assertEquals(1, actionCount);
+    }
+  }
+
+  @Test
+  public void testInterruptSleep() throws Exception {
+    // interrupt the current thread so as to cause the retry's sleep() to throw
+    // an InterruptedException
+    Thread.currentThread().interrupt();
+    boolean result = Retry.action(unsuccessfulAction).sleep(10, TimeUnit.SECONDS).run();
+    Assert.assertFalse(result);
+    Assert.assertEquals(1, actionCount);
+    // This thread should be marked as interrupted (plus clear the flag for the test)
+    Assert.assertTrue(Thread.interrupted());
+  }
+
+  @Test
+  public void testInvalid_maximumRetries() {
+    try {
+      Retry.action(successfulAction).maximumRetries(0);
+      Assert.fail();
+    } catch (IllegalArgumentException ex) {
+      /* expected */
```

Comment:
nit: I think it's meaningful to add a helpful message like the value should be non-negative. It also makes it clear that the exception should come from `.maximumRetries()` and not `.action()`.